### PR TITLE
Smarter Transition From Syncing To Steady State PoS Consensus

### PR DESCRIPTION
**Pull request by tholonious**

### CODE REVIEW

**Comment by tholonious** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1088?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1088** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1088?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1087** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1087?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `feature/pos-networking-and-syncing`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tholonious and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/pos_server_regtest.go [hunk 1]
```diff
@@ -90,17 +90,3 @@ func (srv *Server) submitRegtestValidatorRegistrationTxns(block *MsgDeSoBlock) {
 		}
 	}
 }
-
-func (srv *Server) startRegtestFastHotStuffConsensus(block *MsgDeSoBlock) {
```
Comment (tholonious) [inline lib/pos_server_regtest.go:94]: This hook was a special case we previously used to start the consensus on regtest. We don't need it anymore now that we have the generalized transition detector introduced by this PR

File: lib/server.go [hunk 2]
```diff
@@ -2939,19 +2981,13 @@ func (srv *Server) Start() {
...
-	// off the PoS consensus once the miner is done.
-	if srv.params.NetworkType == NetworkType_TESTNET && srv.fastHotStuffConsensus != nil {
-		tipHeight := uint64(srv.blockchain.blockTip().Height)
-		if srv.params.IsFinalPoWBlockHeight(tipHeight) || srv.params.IsPoSBlockHeight(tipHeight) {
-			if err := srv.fastHotStuffConsensus.Start(); err != nil {
-				glog.Errorf("NewServer: Error starting fast hotstuff consensus %v", err)
-			}
-		}
+	// Initialize state syncer mempool job, if needed.
+	if srv.stateChangeSyncer != nil {
```
Comment (tholonious) [inline lib/server.go:2985]: @superzordon what's the best place to trigger the state syncer mempool sync routine? With PoS, it's possible for a node to start up, no sync blocks from a peer, and instead join the validator set directly.

It seems cleanest to decouple this mempool sync routine from the _startSync operation where it originally was. I have no strong opinion on this, so LMK if this is an approprate place to move this to

File: lib/server.go [hunk 3]
```diff
@@ -2813,13 +2798,70 @@ func (srv *Server) _startAddressRelayer() {
...
+	// If the FastHotStuffConsensus is already running, then there is nothing to do.
+	if srv.fastHotStuffConsensus.IsRunning() {
+		return
+	}
+
+	// Exit early if the current tip height is below the final PoW block's height. We are ready to
+	// enable the FastHotStuffConsensus once we reach the final block of the PoW protocol. The
+	// FastHotStuffConsensus can only be enabled once it's at or past the final block height of
+	// the PoW protocol.
+	tipHeight := uint64(srv.blockchain.blockTip().Height)
```
Comment (lazynina) [inline lib/server.go:2839]: Do we need a chain lock to read this?

File: lib/server.go (hunk 2 above)
Comment (lazynina) [reply lib/server.go:2985]: It belongs in server.Start - putting it in sync is incorrect and should already be addressed in main (or Z has a PR for it).

**Review by lazynina — CHANGES_REQUESTED** [review]

File: lib/server.go [hunk 4]
```diff
@@ -1708,12 +1707,6 @@ func (srv *Server) _startSync() {
 		"header tip height %v from peer %v", bestHeight, bestPeer)
 
 	srv.SyncPeer = bestPeer
-
-	// Initialize state syncer mempool job, if needed.
```
[the remark this answers is not in the document]
Comment (tholonious) [reply lib/server.go:1712]: Marking as resolved since now that Zordon has added this change in main

File: lib/server.go (hunk 3 above)
Comment (tholonious) [reply lib/server.go:2839]: Yes, it's good to hold the read lock for completeness

**Review by lazynina — APPROVED** [review]

**Comment by tholonious** [comment]
### Merge activity

* **Mar 6, 10:34 AM EST**: @tholonious started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/1088).

**Pull request merged**